packages/block-library/src/footnotes/edit.js

```diff
@@ -0,0 +1,127 @@
+/**
+ * WordPress dependencies
+ */
+import {
+	RichText,
+	useBlockProps,
+	getRichTextValues,
+} from '@wordpress/block-editor';
+import { addFilter } from '@wordpress/hooks';
+import { createBlock } from '@wordpress/blocks';
+import { useRefEffect } from '@wordpress/compose';
+
+addFilter(
+	'blockEditor.higherOrderReducer',
+	'core/footnotes',
+	( state, reducer ) => {
+		const blocks = state.tree.get( '' )?.innerBlocks;
+
+		if ( ! blocks ) return state;
+
+		const content = getRichTextValues( blocks ).join( '' );
+
+		if ( ! content ) return state;
+		if ( content.indexOf( 'data-fn' ) === -1 ) return state;
+
+		// This can be avoided when
+		// https://github.com/WordPress/gutenberg/pull/43204 lands. We can then
+		// get the order directly from the rich text values.
+		const regex = /data-fn="([^"]+)"/g;
+		const newOrder = [];
+		let match;
+		while ( ( match = regex.exec( content ) ) !== null ) {
+			newOrder.push( match[ 1 ] );
+		}
+
+		const fnBlock = Array.from( state.byClientId.values() ).find(
+			( block ) => block.name === 'core/footnotes'
+		);
+
+		const currentOrder = fnBlock
+			? state.attributes
+					.get( fnBlock.clientId )
+					.footnotes.map( ( fn ) => fn.id )
+			: [];
+
+		if ( currentOrder.join( '' ) === newOrder.join( '' ) ) return state;
+
+		if ( ! fnBlock ) {
+			const rootBlock = Array.from( state.byClientId.values() ).find(
+				( block ) => block.name === 'core/post-content'
+			);
+			return reducer( state, {
+				type: 'INSERT_BLOCKS',
```

Comment:
I'd be great if the block has access to the raw actions.

Reply:
I think blocks should never have access to state format or raw actions because these changes we evolve often for performance reasons... That said, for this specific case, it could be a private API.